core/src/test/scala/unit/kafka/admin/LeaderElectionCommandTest.scala

```diff
@@ -55,9 +54,18 @@ final class LeaderElectionCommandTest(cluster: ClusterInstance) {
     clusterConfig.serverProperties().put(KafkaConfig.OffsetsTopicReplicationFactorProp, "2")
   }
 
+  def waitForAdminClientHaveNumBrokers(numBrokers: Int): Admin = {
+    // Use a temporary adminClient to wait for all brokers up
+    // If we don't wait for the brokers up, we might have race condition that the metadata in adminClient only has broker2 or broker3 up,
+    // and after broker2/broker3 shutdown, no brokers are available to connect, which causes request timeout
+    TestUtils.waitForNumNodesUp(cluster.createAdminClient(), numBrokers)
+
+    cluster.createAdminClient()
+  }
+
   @ClusterTest
   def testAllTopicPartition(): Unit = {
-    val client = cluster.createAdminClient()
+    val client = waitForAdminClientHaveNumBrokers(3)
```

Comment:
At L73, there is `cluster.waitForReadyBrokers()` which wait until the cluster is ready. Would it make sense to move the creating of the admin client after that in all tests?

Reply:
+1 This makes sense to me.

Reply:
@hachikuji , it doesn't work to put the admin client creating after `cluster.waitForReadyBrokers()`, because the `waitForReadyBrokers` only wait for all brokers registered and unfenced (ref: [here](https://github.com/apache/kafka/blob/trunk/core/src/test/java/kafka/testkit/KafkaClusterTestKit.java#L375)), but the metadata cache in broker might have not updated the cluster metadata. I think we need to explicitly wait for all brokers up via describeCluster and then create adminClient for testing.

Reply:
Ah, I didn't see @showuon's comment below about metadata cache consistency. I guess the issue is that unfencing does not necessarily imply that the broker is caught up to the end of the metadata log. I know previously we were considering having the broker consume its own registration before it unfenced, but it looks like we haven't implemented that. Perhaps a short-term fix is to change `KafkaClusterTestKit.waitForReadyBrokers` to check metadata caches directly?

Reply:
> Perhaps a short-term fix is to change KafkaClusterTestKit.waitForReadyBrokers to check metadata caches directly?

Sounds good. Let me give it a try. Thanks for the comment.